test/extensions/filters/network/thrift_proxy/conn_manager_test.cc

```diff
@@ -1058,6 +1085,178 @@ TEST_F(ThriftConnectionManagerTest, ResetDownstreamConnection) {
+  EXPECT_EQ(0U, store_.counter("test.response_error").value());
+}
+
+// Test the case where there is a limit on the number of requests and the actual number of requests
+// happens to reach the limit.
+TEST_F(ThriftConnectionManagerTest, RequestWithMaxRequestsLimitAndReached) {
+  const std::string yaml = R"EOF(
+    stat_prefix: test
+    route_config:
+      name: local_route
+    max_requests_per_connection: 50
+    )EOF";
+
+  initializeFilter(yaml);
+  EXPECT_EQ(50, config_->maxRequestsPerConnection());
+
+  // Since max requests per connection is set to 50, the connection will be disconnected after
+  // all 50 requests is completed.
+  EXPECT_CALL(filter_callbacks_.connection_, close(Network::ConnectionCloseType::FlushWrite));
+
+  EXPECT_EQ(50, sendSomeThriftRequest(50));
+
+  filter_callbacks_.connection_.dispatcher_.clearDeferredDeleteList();
+
+  EXPECT_EQ(1U, store_.counter("test.downstream_cx_max_requests").value());
+  EXPECT_EQ(50U, store_.counter("test.request").value());
+  EXPECT_EQ(50U, store_.counter("test.request_call").value());
+  EXPECT_EQ(0U, stats_.request_active_.value());
+  EXPECT_EQ(50U, store_.counter("test.response").value());
+  EXPECT_EQ(50U, store_.counter("test.response_reply").value());
+  EXPECT_EQ(0U, store_.counter("test.response_exception").value());
+  EXPECT_EQ(0U, store_.counter("test.response_invalid_type").value());
+  EXPECT_EQ(50U, store_.counter("test.response_success").value());
+  EXPECT_EQ(0U, store_.counter("test.response_error").value());
+}
+
+// Test the case where there is a limit on the number of requests and the actual number of requests
+// exceeds the limit.
+TEST_F(ThriftConnectionManagerTest, RequestWithMaxRequestsLimitAndReachedWithMoreRequests) {
+  const std::string yaml = R"EOF(
+    stat_prefix: test
+    route_config:
+      name: local_route
+    max_requests_per_connection: 50
+    )EOF";
+
+  initializeFilter(yaml);
+  EXPECT_EQ(50, config_->maxRequestsPerConnection());
+
+  // Since max requests per connection is set to 50, the connection will be disconnected after
+  // all 50 requests is completed.
+  EXPECT_CALL(filter_callbacks_.connection_, close(Network::ConnectionCloseType::FlushWrite));
+
+  EXPECT_EQ(50, sendSomeThriftRequest(55));
+
+  filter_callbacks_.connection_.dispatcher_.clearDeferredDeleteList();
+
+  EXPECT_EQ(1U, store_.counter("test.downstream_cx_max_requests").value());
+  EXPECT_EQ(50U, store_.counter("test.request").value());
+  EXPECT_EQ(50U, store_.counter("test.request_call").value());
+  EXPECT_EQ(0U, stats_.request_active_.value());
+  EXPECT_EQ(50U, store_.counter("test.response").value());
+  EXPECT_EQ(50U, store_.counter("test.response_reply").value());
+  EXPECT_EQ(0U, store_.counter("test.response_exception").value());
+  EXPECT_EQ(0U, store_.counter("test.response_invalid_type").value());
+  EXPECT_EQ(50U, store_.counter("test.response_success").value());
+  EXPECT_EQ(0U, store_.counter("test.response_error").value());
```

Comment:
should we check here if `cx_destroy_local_with_active_rq` was incremented? I think it probably is incremented when a close happens after max requests is reached.

Reply:
I don't think cx_destroy_local_with_active_rq will increase because now our requests are sent and get responses sequentially. When the connection is disconnected, there will be no requests in pending. But I can try to construct a scenario like this to test that cx_destroy_local_with_active_rq increases.